runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java

```diff
@@ -364,5 +366,17 @@ public State awaitCompletion() throws Throwable {
       }
       return state;
     }
+
+    @Override
+    public State cancel() throws IOException {
+      throw new UnsupportedOperationException("DirectPipelineResult does not support cancel.");
+    }
+
+    @Override
+    public State waitToFinish(Duration duration)
+        throws IOException, InterruptedException {
+      throw new UnsupportedOperationException(
+          "DirectPipelineResult does not support waitToFinish.");
```

Comment:
DirectRunner.run() can be non blocking.
This function will be awaitCompletion() with the support of duration.